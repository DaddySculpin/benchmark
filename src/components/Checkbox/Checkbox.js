import React from 'react';
import { useId } from '@reach/auto-id';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { Box } from '../Styled/Base';
import VisuallyHidden from '../Styled/VisuallyHidden';
import Icon from '../Icon/Icon.js';
import css from './Checkbox.module.css';

function Checkbox({
  id,
  checked = false,
  disabled = false,
  itemId,
  value,
  onBlur,
  onClick,
  onChange,
  onFocus
}) {
  const inputProps = {
    checked,
    disabled,
    onClick,
    onChange,
    onFocus,
    onBlur,
    value
  };

  const classes = classnames(css.checkbox, {
    [css.disabled]: disabled
  });
  const inputId = `checkbox-${useId()}`;
  const icon = checked ? 'checkbox-selected' : 'checkbox-unselected';

  return (
    <Box
      id={id}
      role="checkbox"
      className={classes}
      aria-checked={checked}
      onClick={onClick}
    >
      <VisuallyHidden>
        <input
          tabIndex="0"
          id={inputId}
          type="checkbox"
          className={css.input}
          {...inputProps}
        />
      </VisuallyHidden>
      <Icon className={css.icon} iconName={icon} size="medium" />
    </Box>
  );
}

Checkbox.propTypes = {
  // dom
  id: PropTypes.string,
  disabled: PropTypes.bool,
  // data
  checked: PropTypes.bool,
  value: PropTypes.string,
  // events
  onFocus: PropTypes.func,
  onBlur: PropTypes.func,
  onClick: PropTypes.func,
  onChange: PropTypes.func
};

export default Checkbox;
