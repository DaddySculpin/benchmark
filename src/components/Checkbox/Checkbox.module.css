.input {
  position: absolute;
  opacity: 0;
}

.checkbox:focus-within {
  outline: 1px dotted;
}

.checkbox {
  display: flex;
  align-items: center;
}

.disabled {
  /* cursor: default; */
}

.disabled .icon {
  opacity: 0.3;
}
