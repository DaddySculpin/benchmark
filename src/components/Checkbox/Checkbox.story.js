import React, { useState } from 'react';
import { storiesOf } from '@storybook/react';

import Checkbox from './Checkbox';

storiesOf('Classic', module).add('Elements: Checkbox', () => {
  const [selected, setSelected] = useState('a');

  function handleChange(e) {
    setSelected(e.target.value);
  }

  return (
    <div role="group">
      <Checkbox
        value="a"
        name="radio"
        checked={selected === 'a'}
        onChange={handleChange}
      >
        Option A
      </Checkbox>
      <Checkbox
        value="b"
        name="radio"
        checked={selected === 'b'}
        onChange={handleChange}
      >
        Option B
      </Checkbox>
      <Checkbox
        disabled
        value="c"
        name="radio"
        checked={selected === 'c'}
        onChange={handleChange}
      >
        Disabled
      </Checkbox>
    </div>
  );
});
