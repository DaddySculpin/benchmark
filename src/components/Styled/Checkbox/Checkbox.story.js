import React from 'react';
import { storiesOf } from '@storybook/react';
import { Global, ThemeProvider } from '../../../index.js';

import { useArrayToggle } from '../../../util/hooks';
import Checkbox from './Checkbox';

storiesOf('Checkbox', module).add('Checkbox', () => {
  const [selected, setSelected] = useArrayToggle(['b']);

  return (
    <Global>
      <ThemeProvider>
        <div role="group">
          <Checkbox
            value="a"
            name="radio"
            checked={selected.includes('a')}
            onChange={value => setSelected(value)}
          >
            Option A
          </Checkbox>
          <Checkbox
            value="b"
            name="radio"
            checked={selected.includes('b')}
            onChange={value => setSelected(value)}
          >
            Option B
          </Checkbox>
          <Checkbox
            disabled
            value="c"
            name="radio"
            checked={selected.includes('c')}
            onChange={value => setSelected(value)}
          >
            Disabled
          </Checkbox>
        </div>
      </ThemeProvider>
    </Global>
  );
});
