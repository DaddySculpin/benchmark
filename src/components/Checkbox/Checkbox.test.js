import React from 'react';
import { axe, toHaveNoViolations } from 'jest-axe';
import { render } from '@testing-library/react';
import Checkbox from './Checkbox';

expect.extend(toHaveNoViolations);

function handleChange() {
  console.log('change');
}

describe('Checkbox', () => {
  it('renders without crashing', () => {
    render(<Checkbox onChange={handleChange} />);
  });

  xit('should not have basic accessibility issues', async () => {
    const { container } = render(<Checkbox onChange={handleChange} />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  xit('should have a label element wrapping the input', () => {
    throw new Error();
  });

  it('should show a radio-selected icon when checked', () => {
    const { container } = render(<Checkbox checked onChange={handleChange} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should show a radio-selected icon when checked', () => {
    const { container } = render(<Checkbox checked onChange={handleChange} />);
    const selected = container.querySelector('.checkbox-selected');
    expect(selected).toBeTruthy();
  });

  it('should show a radio-unselected icon when not checked', () => {
    const { container } = render(<Checkbox onChange={handleChange} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  xit('should change the icon opacity when disabled', () => {
    throw new Error();
  });

  xit('should not change the text opacity when disabled', () => {
    throw new Error();
  });

  xit('should not change the text opacity when disabled', () => {
    throw new Error();
  });

  xit('should show a focus ring when focused', () => {
    throw new Error();
  });

  xit('should not show a focus ring when unfocused', () => {
    throw new Error();
  });

  xit('should invoke the change handler', () => {
    throw new Error();
  });

  xit('should invoke the focus handler', () => {
    throw new Error();
  });

  xit('should invoke the blur handler', () => {
    throw new Error();
  });
});
