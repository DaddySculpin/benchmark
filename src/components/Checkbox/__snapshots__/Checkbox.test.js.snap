// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox should show a radio-selected icon when checked 1`] = `
<div
  aria-checked="true"
  class="checkbox css-vurnku"
  role="checkbox"
>
  <span
    style="position: absolute; height: 1px; width: 1px; clip-path: polygon(0px 0px, 0px 0px, 0px 0px); overflow: hidden !important; white-space: nowrap;"
  >
    <input
      checked=""
      class="input"
      id="checkbox-3"
      tabindex="0"
      type="checkbox"
      value=""
    />
  </span>
  <span
    class="icon icon checkbox-selected medium"
    data-icon=""
  />
</div>
`;

exports[`Checkbox should show a radio-unselected icon when not checked 1`] = `
<div
  aria-checked="false"
  class="checkbox css-vurnku"
  role="checkbox"
>
  <span
    style="position: absolute; height: 1px; width: 1px; clip-path: polygon(0px 0px, 0px 0px, 0px 0px); overflow: hidden !important; white-space: nowrap;"
  >
    <input
      class="input"
      id="checkbox-5"
      tabindex="0"
      type="checkbox"
      value=""
    />
  </span>
  <span
    class="icon icon checkbox-unselected medium"
    data-icon=""
  />
</div>
`;
